### ISSUE TITLE
Luminary Autopilot bug

### ISSUE THREAD

**Issue by spacex15** [rburkey2005/virtualagc#30]
when doing a v49 maneuver with a pitch angle greater than 180 deg
(v49E, V25E,+00000E,+33000E,+00000E,PRO,PRO) for example

the autopilot is executing the correct maneuver but will not maintain the attitude correctly

it slowly drifts toward zero pitch attitude (very slowly, several minutes)

telemetry is showing that variable CDUYD value is decreasing slowly towards zero instead of not moving.

code not working is in P-AXIS_RCS_AUTOPILOT
1STOTWOS	CCS	 A
		AD	ONE
		TC	Q
		CS	A
		TC	Q

this function, instead of storing the CDUYD variable without changing it , is  substracting one from the variable before returning to the caller

it seems there is an implementation problem in the MSU operand when calculating on CDU counters greater than 16384

**Comment by indy91** [comment]
I have stumbled upon the same issue. Any insight in how to fix it?

**Comment by rburkey2005** [comment]
I don't have any real insight into it in the short term.

My fear is that it's a low-level bug in the CPU simulator, and that I my not personally have the expertise to fix it right now.  Part of the problem is that the CPU validation software we have right now is based on _descriptions_ of tests, from a testing document that may have been obsolete with respect to Block 2.

What I'm hoping in the longer term is this:  There is a possibility that in the relatively near future we might have a listing of a program called AURORA which, I'm told, is the _last_ version of the Luminary software to actually contain the official Block 2 version of the testing code.  Once we have that, we can obviously run those tests, and may be able to find and fix specific low-level problems in the CPU.

Technically, the problem is that the behaviors of the CPU's instructions were supposed to be dependent on microcode that the AGC developers called "pulse sequences", but I find those pulse sequences very confusing and got nowhere trying to use those as a guide to implementation, instead basing the behavior on the documented behavior of the instructions.  Some people _don't_ seem to find the pulse sequences confusing, and if there were a programmer who understood the pulse sequences better than I do, it might be possible to track down problems like this prior to having AURORA in hand.

The bottom line is that while I'll still think about it, this problem may surpass my meager programming abilities until AURORA becomes available.

**Comment by rburkey2005** [comment]
There's one thing you might try to help isolate the problem, though I don't really promise it and don't know if it's feasible for you.  If you have any way to download and run Julian Webb's Windows XP LGC simulator (http://www.cems.uwe.ac.uk/~jtwebb/agc/), and to see if it has the same problem, it might be interesting.  I know that Julian actually did work from the pulse sequences, and I know that his CPU simulator does have some slight behavioral discrepancies compared to mine.  I used to think of those differences as flaws in his simulator, but it's entirely possible they're flaws in mine!

**Comment by indy91** [comment]
I came across this issue while working on the Lunar Module in NASSP (With the right padload I got the P63 ignition algorithm to work, yay!). I will try the LGC simulator and look, if the behaviour is the same.

**Comment by rburkey2005** [comment]
My brain has started working a little bit, and I have recalled that there's something special about the CDU counters.  Unlike everything else in the AGC, they used a 2's complement numerical format rather than a 1's complement numerical format used for everything else.  If NASSP had mistakenly used a 1's complement for everything (including the CDU's), then the counters would operate normally in the range 0-16383, but would function abnormally at 16384 or above, such as counting in the wrong direction.  Is there a possibility that NASSP has implemented this wrong?

**Comment by indy91** [comment]
It's very much possible there is something wrong in NASSP. It all works fine with the CM AGCs though, no problem holding attitude in Colossus249, Comanche055 or Artemis072. Could be something with the implementation specific to the LM of course. But I am not very knowledgable about the CDUs etc, I usually treat the Virtual AGC implementation as the real thing and work on NASSP with that assumption.

**Comment by thewonderidiot** [comment]
So, my hardware simulator actually failed one of validation's MSU checks. I analyzed the control pulses the MSU0 sequence was generating, and it looked in line with what Hugh Blair-Smith listed in AGC4 Memo #9. In the end, I couldn't find anything wrong with what the simulation was doing, so I changed my copy of validation locally to match the answer I was getting.

So I think @indy91 is exactly write, and there's an implementation problem with MSU. Of course, VirtualAGC was coded to match what's specified in E-2065, so if we're right than either the document is wrong or the associated flowchart got overwritten by bad OCR.

Either way, AURORA should hopefully be able to settle this once and for all. I can throw together a version of MSU that matches my hardware simulation in the meantime.

**Comment by thewonderidiot** [comment]
The specific test I think E-2065 has wrong is this one: 
![msu_error](https://cloud.githubusercontent.com/assets/94056/17835268/bed2557e-6719-11e6-87d7-60c2365cebd3.png)

which was recreated [here](https://github.com/rburkey2005/virtualagc/blob/c56c9cdce8ab963441a333ce728d832e85a6b229/Validation/SmallyMSUCHK.agc#L43) in Validation.

I get +1 in my simulator rather than +0. And I think an error here is believable, considering the next box in the flowchart subtracts a totally bogus 400000. :) I'll include a detailed analysis of MSU's behavior here with my pull request.

**Comment by rburkey2005** [comment]
Well, I've personally come to suspect that yaAGC does have a problem 
with respect to some pulse sequences, but that's largely because I know 
that Julian Webb's LGC simulator fails several parts of the Validation 
program, and I know that Julian payed close attention to the pulse 
sequences when designing his sim.  Unfortunately, his code isn't open, 
and I didn't get any where trying to cooperate with him in figuring out 
the right answer, so I've never been able to figure out for sure which 
one (if either) is correct.  My fear is that, like so much of the Apollo 
documentation, the tests published by Smally had become out of date, and 
were thus wrong in subtle ways.  So if there's a "final" "official" 
version of those tests in AURORA, it will certainly be swell.

**Comment by rburkey2005** [comment]
@thewonderidiot As you may suspect, my last comment was actually written before I saw _your_ last comment.  But remind me, does running the Validation fail in other ways on the hardware sim?  I was vaguely toying with asking whether you wanted to use the same method to check the other Validation discrepancies between VirtualAGC and Julian Webb's LGC sim, but if that MSU thing was the last issue you had with it, there wouldn't be anything left to check, would there?

**Comment by thewonderidiot** [comment]
@rburkey2005 Yeah, there's a few more discrepancies between the hardware sim and VirtualAGC that I've found:

1. Only 15 bits can be read from the L register, so it can't store an overflow condition. Validation checks that it can. I've been on the fence for a while about whether who was right about this one, but yesterday I found this in E-2065: "D--LCHK   This subroutine  was  written to check that the overflow bit disappeared when a  word went into and out of the L register and to make sure that the  Q  register was capable of holding 16  bits"

2. All of the I/O channels in the AGC are flip-flops, and in general if a channel only has N bits, there will only be N flip-flops present. Validation uses channels 26 and 27, which aren't real, so it fails on hardware. I switched to 11 and 12 for all those tests, since are both 15 bits long and both readable and writeable. It's not worth changing yaAGC here, since it uses fake channels to communicate with other things.

3. For DV, when the divisor and dividend are equal in magnitude but have opposite signs, we get different remainders. VirtualAGC gives the value of the divisor, which agrees with E-2065. I get the value of the dividend, which agrees with E-2052: "If a quantity  is divided into a quantity  of equal magnitude, we get a quotient of either posmax or negmax and a remainder equal to the dividend." This one's a toss-up, and I've been on the fence about it for a while, but at this point I'm slightly more inclined to believe E-2052.

If Julian Webb's sim was built around control pulses, *#*3 would be the only one of these that he might find. *That being said*, I've found that there are a few instructions for which the control pulses don't tell the whole story -- some instructions can have side effects or behaviors not captured by the pulses. DV happens to be one of those instructions. In the special case of the Validation check mentioned above, those side effects don't happen, and so the control pulses tell the whole story. However it wouldn't surprise me if other cases where DV does get more weird would get different answers in a control-pulse-based sim.

**Comment by thewonderidiot** [comment]
Hmm.... do you happen to know *how* to go about running Validation on Julian Webb's simulator?

**Comment by rburkey2005** [comment]
Yes, there's a utility called webb2rburkey-rope.c in the Tools directory 
that converts the (non-hardware :-) ) VirtualAGC rope format to Julian's 
format, and then you have to find his rope file (which I think is 
Luminary131, but I don't know the filename) and simply replace it with 
the new one.

**Comment by rburkey2005** [comment]
I think my reasoning as explained in ValidateIO.agc for using unreal 
channels was sound enough as far as it goes, but I guess I overlooked 
the notion of running it on real hardware.  I don't have a problem with 
changing it to use real channels, as long as they're "safe" in the sense 
of having no unintended consequences.  In fact, I think it would be 
better to do so, though I'm not adamant.

For the others, they're both taken from Smally, so hopefully they're in 
AURORA, and it doesn't sound like you feel strongly enough about them to 
for us to do anything about them prior to seeing what AURORA does with 
them.  Would that be a fair assessment?

**Comment by thewonderidiot** [comment]
Perfect, thanks!
And, wow, his simulator gets a lot of errors. It's hard to guess what's going on without seeing the source for it.

*#*1,0: I guess it's finding the wrong thing for Z? VirtualAGC is right here
*#*4,2: CCS failure. No idea what went wrong here, but hardware sim and VirtualAGC agree
*#*6,2: DAS failure. Again don't know, but hardware sim and VirtualAGC agree
*#*11,10: LXCH error, possibly related to losing overflow. However, more than just this one should fail, and they don't.
*#*12,04: INCR error, when using L. I don't think overflow matters here? I didn't have to change the INCR tests for the hardware sim
*#*13,01: ADS error. Doesn't happen in VirtualAGC or hardware sim
*#*15,02: CS error when using L. Doesn't happen in VirtualAGC or hardware sim
*#*16,05: TS error because of L overflow. This one is real, and hardware sim agrees with it. The LGC simulator totally misses 16,10 though, which should also fail.
*#*25,07: XCH error because of L overflow. Also real, hardware sim agrees.
*#*26,04: AD error because of L overflow. Also real, hardware sim agrees.
*#*30,06: RAND failure... possibly because it hits L? VirtualAGC and hardware sim don't show this
*#*61,37: DAS failure? Not sure why. Hardware sim doesn't show this
*#*61,42: Another DAS failure? Really not sure. Interestingly, both 61,37 and this one are places where you disagreed with Smally.
*#*61,44: Failure when fiddling L? I think L handling in general seems a bit sketchy
*#*61,45: More possible L weirdness
*#*61,50: No idea on this one either.
*#*61,54: Division failure, although not the one the hardware sim sees. Probably due to DV weirdness

There's several more after that but I'm gonna stop there. I agree with your original position, it's probably best to assume VirtualAGC is correct on most of these differences.

**Comment by thewonderidiot** [comment]
Yeah, I've largely been running Validation without a DSKY since it's not easy to loop one into the verilog simulator -- so I got a nice surprise of CH11 making the DSKY go wild during the I/O checks. I'll review the channels again to see if there's a nicer one to use.

The DV thing we can probably leave until the arrival of AURORA to change. It's an easy fix in yaAGC.

For the L register size, it might be worth getting a head start. I've been looking through agc_engine.c and it doesn't look like there's an easy place to drop in special behavior of L reads. The behavior of the hardware is:
- Writing a 16-bit value to L puts all 16 bits into the L flip-flops
- Reading from L ignores the stored value of bit 15, instead copying bit 16 into the bit 15 location, effectively doing an overflow correction

I'm not sure if it would be easier to do the overflow correction on write or on read, but it doesn't matter since it's all invisible to software. Which do you think would be easier to implement?

**Comment by rburkey2005** [comment]
I don't know what would be easier.  I would probably do it every place 
the simulated register gets written to.  Why save an incorrect bit for 
any period of time whatever?

**Comment by thewonderidiot** [comment]
Fair point -- I had been thinking that maybe the Monitor could see the full 16 bits of L, but reviewing the monitor I/O lines I'm not sure of that any more. I think it just snoops for times when L is moving across the write bus for its display, so it'll only ever see the overflow-corrected version too.

I'll start working on changes to agc_engine.c that overflow-corrects L on write.

**Comment by rburkey2005** [comment]
Well, my original assumption may have been mostly my own hubris. :-)

And I guess it was also based on my belief (though I don't know that 
Julian said anything to make me think so) that while VirtualAGC was 
designed from the bottom up, I thought Julian's AGC was designed from 
the top down.  I.e., I wanted the simulated CPU to work exactly 
according to the documentation and wanted to drop Luminary and Colossus 
into it and have them just magically work, without my necessarily 
understanding what they were supposed to do, whereas Julian knew what he 
wanted the end product to be (Luminary), and then implemented the 
underlying machinery until he got the results he expected.  That was the 
explanation I supplied in my mind as to why he was indifferent when I 
presented the validation discrepancies to him.  And it's a natural 
reaction:  To him they couldn't be serious problems if he was getting 
the results he wanted from Luminary, whereas to me they were serious, 
since they meant the underlying structure wasn't correct, regardless of 
what happened in any given version of the AGC code.

For example, my recollection is that Colossus failed immediately on his sim.

But still, it doesn't make me feel any better about ignoring the 
results.  But you're also right, in that I still don't know what, if 
anything can be done about it!

**Comment by indy91** [comment]
We are only really just now getting into using Luminary for NASSP. So far all the Colossus versions have been working beautifully without problem and there is really not much left that hasn't been tried yet. If we are seeing problems like this DAP issue that are noticeable for the "end user", then we will discover more of them soon, I guess.

**Comment by rburkey2005** [comment]
Cool! I'd prefer to keep working on yaYUL right now anyway. Please make it easy to undo, though, in case it ends up breaking nassp or something.


Sent on my Virgin Mobile Samsung Galaxy S® 5

-------- Original message --------

**Comment by rburkey2005** [comment]
That's good to know. As a software developer, we always like to assume that no problems being reported means no problems at all, so it's good to know that that's at least a fair approximation in this case. :-)


Sent on my Virgin Mobile Samsung Galaxy S® 5

-------- Original message --------

**Comment by thewonderidiot** [comment]
I'll do my best, but there's a lot of instructions that hit c (RegL) directly, plus direct addressing at address 1 and channel 1 (or is L channel 2?). Considering the myriad of ways to mutate L, it probably wouldn't be a bad idea to expand Validation to test all of the venues. AURORA is probably only going to try one instruction since there'a only one way for the hardware to read L (RL control pulse).

We can just leave the PR open if I have time to complete it before AURORA shows up. And in the worst case, there's always git revert :)

**Comment by rburkey2005** [comment]
Okay. No problem.


Sent on my Virgin Mobile Samsung Galaxy S® 5

-------- Original message --------

**Comment by rburkey2005** [comment]
Much, much (trivially) easier way to do it.  Let the reads and writes to 
L alone.  At the start of agc_engine(), simply zero out bit 16 of L.  So 
the bit gets written to L, but is forgotten again before anything can be 
done with it later.  Can even make a command-line option to enable or 
disable that feature.

**Comment by thewonderidiot** [comment]
Ah, that's an excellent idea! In what situation might we want it to be switchable though? It's either (probably) correct hardware behavior, or it's not.

**Comment by rburkey2005** [comment]
I can't say I have anything in mind. It's just a conservative thing to do if there's any doubt whatsoever, and it's very cheap. But you're probably right that it would never need to be used, and it would be easy enough to implement later if needed.
It has also occurred to me that the best place to do it is probably not at the top of agc_engine, but rather as soon as possible in agc_engine after the simulated instructions are executed, just in case there's anything else within agc_engine itself that wants to use the register contents.


Sent on my Virgin Mobile Samsung Galaxy S® 5

-------- Original message --------

**Comment by thewonderidiot** [comment]
Roger that. Down in where Z is updated and the bank registers are mirrored looks like a good place :)
